### DISCUSSION TITLE
Implement DID Connect for Decentralized Identity Authentication in Authelia

### DISCUSSION THREAD

**Discussion by Tim-Rambo** [authelia/authelia#8158] [Ideas]
### Description

**Description:**

This feature request proposes integrating Decentralized Identifiers (DIDs) and the DID Connect protocol into Authelia for user authentication. This would allow users to leverage their self-managed digital identities for secure access to resources protected by Authelia.

### Use Case

**Use Case:**

Currently, Authelia primarily relies on traditional authentication methods (e.g., LDAP, OIDC). Integrating DID Connect would enable:

* **User-centric identity management:** Users control their own identities and can selectively disclose information.
* **Enhanced privacy:** Reduced reliance on centralized identity providers.
* **Improved security:** Cryptographic verification of DIDs enhances trust and reduces the risk of identity theft.
* **Seamless authentication:** Streamlined login experiences with user-controlled identities.
* **Future-proof solution:** Aligns with the growing trend of decentralized identity solutions.

### Details

**Details:**

To implement DID Connect support in Authelia, the following functionalities would be required:

* **DID Registration:** Allow users to register their DIDs with Authelia, potentially linking them to existing accounts.
* **DID Connect Integration:** Implement the DID Connect protocol for authentication flows, including request generation, redirection, and response handling.
* **Cryptographic Verification:** Verify the authenticity and integrity of DIDs and associated data using appropriate cryptographic libraries.
* **User Interface:** Provide a user-friendly interface for managing DIDs and initiating DID Connect authentication.

**Alternatives Considered:**

* **Other Decentralized Identity Solutions:**  While alternatives like Verifiable Credentials exist, DID Connect offers a simpler and more focused approach for authentication.
* **Custom Integration:** Building a custom integration with a specific DID provider is possible but less flexible and maintainable compared to native DID Connect support.

**Additional Considerations:**

* **Standardization:** Adhere to DID Connect specifications and best practices to ensure interoperability with various DID methods and providers.
* **Security:** Implement robust security measures to protect user data and prevent unauthorized access.
* **User Experience:** Design a clear and intuitive user experience for DID-based authentication.

By implementing DID Connect, Authelia can provide a modern, privacy-preserving, and user-centric authentication solution that empowers users to control their digital identities.

### Documentation

* **ArcBlock DID Spec:** [https://github.com/ArcBlock/abt-did-spec](https://github.com/ArcBlock/abt-did-spec)
* **ArcBlock:** [https://www.arcblock.io/en](https://www.arcblock.io/en)
* **DID Wallet:** [https://www.didwallet.io/en?locale=en](https://www.didwallet.io/en?locale=en)
* **DID Connect App:** [https://app.didconnect.io/en](https://app.didconnect.io/en)
* **DID Connect Introduction:** [https://www.arcblock.io/docs/did-connect/en/did-connect-introduction](https://www.arcblock.io/docs/did-connect/en/did-connect-introduction)
* **DID Connect Workflow:** [https://www.arcblock.io/docs/did-connect/en/did-connect-workflow](https://www.arcblock.io/docs/did-connect/en/did-connect-workflow)
* **Detailed Explanation of DID Connect:** [https://www.arcblock.io/docs/did-connect/en/did-connect-protocol](https://www.arcblock.io/docs/did-connect/en/did-connect-protocol)

These links provide comprehensive information about the DID Connect protocol, its workflow, and related tools. The detailed explanation offers in-depth knowledge about the protocol's inner workings, including request and response formats, verifiable claims, and state machine perspectives.  The ABT DID specification provides the foundational details for understanding how DIDs are created, managed, and used within the ArcBlock ecosystem.

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)

- [X] I have checked for related issues and checked the documentation

**Comment by james-d-elliott** [comment]
We have some interest in this but want to gauge community interest too. Also we would require a supporting go library in a good state to exist before we adopted something like this. As such we're converting it to an idea discussion where it can be voted on. We'd appreciate that there is actual discussion about this rather then blindly using AI, as well as documentation that is more generic rather than focused on one DID method.

**Reply by james-d-elliott** [reply]
Couple of side notes I forgot to mention, we are interested in other implementations of this to gauge the exact level of support the industry has and the maturity of this technology. At such a time as we can see all of these things exist within this ecosystem we will consider opening this as a formal feature request again.

**Reply by Tim-Rambo** [reply]
Hey, just wanted to add a comment on this, it is an open protocol just like email is, and it has nothing to do with AI. Yes, Arcblock has a lot of other tools in their arsenal which uses AI, but their open protocols that I am talking about does not use any AI whatsoever.

**Reply by james-d-elliott** [reply]
The feature request was AI generated.

**Reply by Tim-Rambo** [reply]
> The feature request was AI generated.

I used AI to help me write the feature request because I am not fluent in English.... I do not see what is so wrong with that?

**Reply by james-d-elliott** [reply]
AI in itself is completely fine, the issue is the lack of clear disclosure. You tried to deflect when it was brought up (sorry there's absolutely no way you can convince me you didn't know what I was talking about, don't bother insulting me by trying), then you try to explain it away. The lack of disclosure is not that bad on its own, but your response when you were confronted comes over very unprofessional.

Either way this is besides the point, we'll need the other points to be addressed to put more effort into this:

1. Discussion here that is absent AI or has clear AI disclosure.
2. Interest from our users.
3. DID Method agnosticism. I do not believe there is any room for particular implementations to define their own specifications that are relevant to Authelia. The inputs and outputs from relevant API's should be the same between methods. As such we are only interested in the area of the spec we need to implement. This exists because we don't want the implementation we do to limit our users due to the whims of any particular party that has a conflict of interest. For example:
  * Generic documentation from standards bodies and divested entities.
  * A go library that performs the necessary specification verification steps from a divested entity (i.e. not related to any particular method).
  * A js library that performs the necessary API steps within the context of the browser if relevant from a divested entity (i.e. not related to any particular method).
4. Information on the relying parties currently using these methods so we can gauge the actual prevalence of the technology.

**Reply by Tim-Rambo** [reply]
> AI in itself is completely fine, the issue is the lack of clear disclosure. You tried to deflect when it was brought up (sorry there's absolutely no way you can convince me you didn't know what I was talking about, don't bother insulting me by trying), then you try to explain it away. The lack of disclosure is not that bad on its own, but your response when you were confronted comes over very unprofessional.

Sorry, I did not mean to come in as unprofessional at all, I misunderstood you completely sorry about that. I will own up to it. Like I said, I am not fluent in English. Sorry again, was not meant in that way at all.

**Reply by james-d-elliott** [reply]
Not a problem. Just want to ensure we have a clear picture of the technology and as you probably know AI can get things wrong. There's nothing wrong with using it as long as we're made aware so we can account for it. This is pretty essential for communication in a professional environment.

**Reply by Tim-Rambo** [reply]
> Not a problem. Just want to ensure we have a clear picture of the technology and as you probably know AI can get things wrong. There's nothing wrong with using it as long as we're made aware so we can account for it. This is pretty essential for communication in a professional environment.

I am still new to the coding and programming world, anyway, how can I help with the implementation of DID Connect for Authelia? I tried to make sense of the 1 to 7 steps with a translator, but did not give me any more insight of what is needed. (Google translator sucks) Could you maybe help me to be able to grasp your 1 to 7 step?

Note: I only used the AI to create the "GitHub Issue/GitHub Feature Request" as I am aware that people can be a bit skeptic about AI code etc.

**Reply by james-d-elliott** [reply]
I've adjusted this, was not meant to be 7 steps. 

Step 1 and 2 will just take time, we need to see if other users are interested in this and their thoughts. Step 3 and 4 however are just information and/or development works. Step 3 will require finding what libraries and documentation exist that isn't related to ABT, step 4 will require finding other applications or services similar to Authelia that already use this.
